keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,569 @@
+
+#### Story 1
+
+As an ML researcher I run AI training workloads which require exchanging huge
+amounts of data between pods. The workloads take twice longer and are twice
+more expensive when there are pods running in different datacenter racks. I
+don't want to start my workload if the pods cannot be placed within the same
+rack.
+
+#### Story 2
+
+As an ML researcher I run AI training workloads which require communicating
+moderate amounts of data between pods. I would like to optimize the runtime of
+the workloads, by placing the pods as close to another as possible, ideally
+within a single rack, but I'm ok running the workload on nodes scattered across
+datacenter if placing all pods within the same hierarchy rack is not possible at
+the given time.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Integration support
+
+In the alpha iteration we aim to support Job & JobSet. If other integrations
+follow naturally we may as well support them in alpha.
+
+##### JobSet
+
+One complication we noticed for JobSet is that the proposed design assumes
+injecting the dedicated scheduling gate, called `kueue.x-k8s.io/topology` into
+the PodTemplate. However, currently JobSet's `spec.replicatedJob` field is
+immutable, even if the JobSet is suspended. This has been relaxed in JobSet 0.6
+(see [PR](https://github.com/kubernetes-sigs/jobset/pull/623)).
+
+#### Support for the "auto" mode
+
+We are considering introducing the "auto" mode where the user does not need to
+specify the level by node label value, but the lowest level is selected
+automatically.
+
+One approach is to accept "auto" as a specific value for the "prefer" and
+"require" labels. However, we are going to defer this for Beta or GA based on
+the users' feedback.
+
+### Risks and Mitigations
+
+#### Non-exclusive use of nodes
+
+The TAS feature assumes exclusive use of the nodes, that is all pods running on
+the nodes come from workloads assigned to the same TAS Resource Flavor. This can
+be achieved by admins by using a distinct label corresponding to a node group
+meant for TAS.
+
+In order to reduce the risk of such misconfiguration we require every TAS
+Resource Flavor to have at least one label.
+
+#### Node topology changes
+
+A node can be removed during runtime of a workload, and the replacement for the
+pod running on this node may not be able to find a replacement matching the
+TopologyAssignment. In that case the workload may not be able to progress.
+
+First, note that the issue also exists for regular workloads, but for TAS
+workloads, given the additional constraints, it might be harder to find a
+replacement node.
+
+In order to mitigate this risk we propose to extend the waitForPodsReady
+mechanism with a new timeout, called replacement timeout, which defines the
+timeout for all pods to be ready again. The time is computed since the last
+transition to the PodsReady=false, more details in the KEP PR. This mechanism
+will also be helpful for regular workloads.
```

Comment:
Actually, I think the current WaitForPodsReady logic could immediately evict a job that transitions to ready=false, so it might need some adjustment too.

Reply:
Right, but I believe it is a detail we can discuss in the KEP, I opened the discussion thread not to forget about it: https://github.com/kubernetes-sigs/kueue/pull/2737/files#r1747008174